File: app/src/main/java/com/example/nasmovie/ui/adapter/FavoriteAdapter.java
package com.example.nasmovie.ui.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.nasmovie.R;
import com.example.nasmovie.data.model.Movie;
import com.example.nasmovie.view.MovieCard;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * 收藏电影适配器
 */
public class FavoriteAdapter extends RecyclerView.Adapter<FavoriteAdapter.ViewHolder> {

    private List<Movie> movies = new ArrayList<>();
    private Set<String> selectedIds = new HashSet<>();
    private boolean selectionMode = false;

    private OnMovieClickListener clickListener;
    private OnMovieLongClickListener longClickListener;

    public interface OnMovieClickListener {
        void onMovieClick(Movie movie, int position);
    }

    public interface OnMovieLongClickListener {
        boolean onMovieLongClick(Movie movie, int position);
    }

    public void setOnMovieClickListener(OnMovieClickListener listener) {
        this.clickListener = listener;
    }

    public void setOnMovieLongClickListener(OnMovieLongClickListener listener) {
        this.longClickListener = listener;
    }

    public void setMovies(List<Movie> movies) {
        this.movies = movies;
        notifyDataSetChanged();
    }

    public List<Movie> getMovies() {
        return movies;
    }

    public void setSelectionMode(boolean enabled) {
        this.selectionMode = enabled;
        if (!enabled) {
            selectedIds.clear();
        }
        notifyDataSetChanged();
    }

    public boolean isSelectionMode() {
        return selectionMode;
    }

    public void toggleSelection(String movieId) {
        if (selectedIds.contains(movieId)) {
            selectedIds.remove(movieId);
        } else {
            selectedIds.add(movieId);
        }
        notifyDataSetChanged();
    }

    public Set<String> getSelectedIds() {
        return selectedIds;
    }

    public int getSelectedCount() {
        return selectedIds.size();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_favorite_movie, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Movie movie = movies.get(position);
        holder.bind(movie, selectedIds.contains(movie.getId()), selectionMode);
    }

    @Override
    public int getItemCount() {
        return movies.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final MovieCard movieCard;

        ViewHolder(View itemView) {
            super(itemView);
            movieCard = itemView.findViewById(R.id.movie_card);

            itemView.setOnClickListener(v -> {
                int position = getBindingAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    Movie movie = movies.get(position);
                    if (selectionMode) {
                        toggleSelection(movie.getId());
                    } else if (clickListener != null) {
                        clickListener.onMovieClick(movie, position);
                    }
                }
            });

            itemView.setOnLongClickListener(v -> {
                int position = getBindingAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    Movie movie = movies.get(position);
                    if (longClickListener != null) {
                        return longClickListener.onMovieLongClick(movie, position);
                    }
                }
                return false;
            });
        }

        void bind(Movie movie, boolean isSelected, boolean selectionMode) {
            movieCard.setMode(MovieCard.CardMode.FAVORITE);
            movieCard.loadMovie(movie);
            movieCard.setSelected(isSelected && selectionMode);
        }
    }
}

File: app/src/main/java/com/example/nasmovie/ui/adapter/FavoriteAdapter.kt
package com.example.nasmovie.ui.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.nasmovie.data.model.Movie
import com.example.nasmovie.databinding.ItemFavoriteMovieBinding
import com.example.nasmovie.view.MovieCard

/**
 * 收藏电影适配器 (Kotlin + ListAdapter)
 */
class FavoriteAdapter : ListAdapter<Movie, FavoriteAdapter.ViewHolder>(MovieDiffCallback()) {

    val selectedIds = mutableSetOf<String>()
    var isSelectionMode = false
        private set

    private var clickListener: OnMovieClickListener? = null
    private var longClickListener: OnMovieLongClickListener? = null

    interface OnMovieClickListener {
        fun onMovieClick(movie: Movie, position: Int)
    }

    interface OnMovieLongClickListener {
        fun onMovieLongClick(movie: Movie, position: Int): Boolean
    }

    fun setOnMovieClickListener(listener: OnMovieClickListener) {
        this.clickListener = listener
    }

    fun setOnMovieLongClickListener(listener: OnMovieLongClickListener) {
        this.longClickListener = listener
    }

    // 为了兼容旧的 Java 代码
    fun setMovies(movies: List<Movie>) {
        submitList(movies)
    }

    fun setSelectionMode(enabled: Boolean) {
        this.isSelectionMode = enabled
        if (!enabled) {
            selectedIds.clear()
        }
        notifyDataSetChanged()
    }

    fun toggleSelection(movieId: String) {
        if (selectedIds.contains(movieId)) {
            selectedIds.remove(movieId)
        } else {
            selectedIds.add(movieId)
        }
        notifyDataSetChanged()
    }

    fun getSelectedCount(): Int {
        return selectedIds.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemFavoriteMovieBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val movie = getItem(position)
        holder.bind(movie, selectedIds.contains(movie.id), isSelectionMode)
    }

    inner class ViewHolder(private val binding: ItemFavoriteMovieBinding) : RecyclerView.ViewHolder(binding.root) {

        init {
            binding.root.setOnClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    val movie = getItem(position)
                    if (isSelectionMode) {
                        toggleSelection(movie.id)
                    } else {
                        clickListener?.onMovieClick(movie, position)
                    }
                }
            }

            binding.root.setOnLongClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    val movie = getItem(position)
                    longClickListener?.onMovieLongClick(movie, position) ?: false
                } else {
                    false
                }
            }
        }

        fun bind(movie: Movie, isSelected: Boolean, selectionMode: Boolean) {
            binding.movieCard.setMode(MovieCard.CardMode.FAVORITE)
            binding.movieCard.loadMovie(movie)
            binding.movieCard.setSelected(isSelected && selectionMode)
        }
    }

    private class MovieDiffCallback : DiffUtil.ItemCallback<Movie>() {
        override fun areItemsTheSame(oldItem: Movie, newItem: Movie): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Movie, newItem: Movie): Boolean {
            return oldItem == newItem
        }
    }
}
